### ISSUE TITLE
Migration from angular's i18n

### ISSUE THREAD

**Issue by sroettering** [ngneat/transloco#139]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[x] Feature request
[ ] Documentation issue or request
[ ] Support request
[ ] Other... Please describe:
</code></pre>

## What is the motivation / use case for changing the behavior?

There is a migration script for ngx-translate but not for angular's built-in i18n.
It would be nice to have one for this as well. Is there anything planned?

**Comment by shaharkazaz** [comment]
@sroettering That's definitely something in our plans 🙂 

The thing is that since Angular's i18n doesn't work on a key-value base we need to create the keys for you, which option you think would be better:
- Hashing the string value into a number and use it as the key (`My sample string` => `8541236`) example for the output JSON:
```
{
  "8541236": "My sample string",
  "4563214": "My title",
  ...
}
```
- use the string as the key by making the entire string kebab case (`My sample string` => `my-sample-string`)  example for the output JSON:
```
{
  "my-sample-string": "My sample string",
  "my-title": "My title",
  ...
}
```

**Comment by sroettering** [comment]
Nice to hear that there are plans!

I would definitely prefer a version where the keys stay readable. Hashing them would make it difficult for translators to understand them.
Also in angular it was possible to use spaces in the keys (well rather in the context) and this does not seem to be supported by transloco, so these would need to be replaced somehow, I think.

Small example from our app:
In our template we have an angular i18n tag like this `i18n="map legend|@@MapLegendLabel"`. We use POEditor for translating the terms and exporting to key_value_json format results in the following:
```
{
  "map legend": {
        "MapLegendLabel": "Legend"
    }
}
```

So at least POEditor treats the context information (text before the "|") as scope but does not remove the spaces.
I guess one way to handle/keep them is to use a custom interceptor, but maybe it would be better to change them during migration right away.

**Comment by shaharkazaz** [comment]
So, if I understand you correctly, the following template:

```html
<h1 i18n>translation value</h1>
<h1 i18n="site header|value 1 sample">Val1</h1>
<h1 i18n="site header|value 2 sample">Val2</h1>
<h1 i18n="other context|another comment@@myId">Val3</h1>
```

Should output the following Transloco JSON:

```json
{
  "translation-value": "translation value",
  "site header": {
     "val1": "Val1",
     "val1.comment": "value 1 sample",
     "val2": "Val2",
     "val2.comment": "value 2 sample",
  },
  "other context": {
     "myId": "Val3",
     "myId.comment": "another comment"
  }
```

Note: the `.comment` is the way we support [comments in Transloco](https://netbasal.gitbook.io/transloco/plugins-1/translators-comments)

**Comment by sroettering** [comment]
Yes that's basically correct in our case.
However I found that using spaces in transloco does not work. So `transloco="site header.val1"` does not work, or I did something wrong.

**Comment by shaharkazaz** [comment]
@sroettering Here is a [working example](https://stackblitz.com/edit/ngneat-transloco-c7oxky)
Are you working with v2?

**Comment by sroettering** [comment]
Nice! That should work for us. I used v2, yes, but I was just evaluating and playing around with the library. We are planning to migrate very soon. It makes it much easier when we can keep the old keys, thanks.

**Comment by shaharkazaz** [comment]
@sroettering Could take a couple of weeks until the script will be ready 🙂

**Comment by NetanelBasal** [comment]
@sroettering Actually I think you can do it pretty quickly with a DOM parser library. Here is a quick code example I made, and it seems to work well:

```ts
const cheerio = require('cheerio');
const glob = require('glob');
const fs = require('fs');

const translation = {};

const htmlFiles = glob.sync(`src/*.html`);
htmlFiles.forEach(path => {
  const html = fs.readFileSync(path, { encoding: 'utf-8' });
  const $ = cheerio.load(html, { decodeEntities: false });
  $('[i18n]').each(function(i, elem) {
    const attr = $(this).attr('i18n');
    const text = $(this).text();
    let context,
      comment,
      key = text;

    if (attr) {
      const value = attr.split('|');
      const splitById = attr.split('@@');
      // we have both context and description
      if (value.length > 1) {
        context = value[0];
        comment = value[1];
        // take the @@id or the text
        key = splitById.length > 1 ? splitById[1] : text;
      } else {
        comment = value[0];
      }
    }

    if (context) {
      translation[context] = translation[context] || {};
      translation[context][key] = text;
      if (comment) {
        translation[context][`${key}.comment`] = comment;
      }
    } else {
      translation[text] = text;
    }

    const resolveKey = context ? `${context}.${text}` : text;
    $(this).text(`{{ '${resolveKey}' | transloco }}`);
    $(this).removeAttr('i18n');
  });

  const result = $('body').html();
  fs.writeFileSync(path, result, { encoding: 'utf-8' });
});

['en', 'es'].forEach(lang => {
  fs.writeFileSync(`${lang}.json`, JSON.stringify(translation, null, 2), { encoding: 'utf-8' });
});
```

**Comment by sroettering** [comment]
@NetanelBasal Thank you for the snippet! I will give it a try soon.

**Comment by sroettering** [comment]
Another thing to keep in mind for the migration are `<ng-container>` elements with translations. These will be rendered as comments and transloco can't do translation work here.
In our apps we will probably replace them with unstyled `<div>`s, but that could maybe not work for other people.

**Comment by NetanelBasal** [comment]
It doesn't do it at runtime. It will transform it to:

`<ng-container>{{ 'key' | transloco }}</ng-container>`

**Comment by NetanelBasal** [comment]
Try the script and let me know.

**Comment by sroettering** [comment]
TLDR; the script does not work out of the box.

On the one hand it does not handle `i18n-<attrName>` tags which is quite easy to fix.
On the other hand cheerio has a lot of issues when handling angular templates for which I found workarounds now. 
I'm still not sure if I covered all the edge cases now. Maybe plain old string replacements would work better for this.

I will share the final version of the script when we finished the migration.

**Comment by NetanelBasal** [comment]
> On the other hand cheerio has a lot of issues when handling angular templates for which I found workarounds now.

Can you provide an example, please?

**Comment by sroettering** [comment]
I had the following problems with cheerio:
* it lower cases attribute names automatically in html mode (ex. [formGroup] -> [formgroup])
  * solution: turn on `xmlMode` and set `lowerCaseAttributeNames` to `false`
* it self closes empty tags (ex. `<div transloco="someKey"></div>` -> `<div transloco="someKey" />`)
  * workaround: insert empty text node for those elements)
* it adds empty quotes to attributes and thus directives (ex. `<tr cdk-row></tr>` -> `<tr cdk-row=""></tr>`)
  * workaround: `const result = $.root().html().replace(/=""/g, '');` remove the empty quotes in the result string
* it can't select elements by having an attribute name starting with something. I need to find all elements with an `i18n` attribute on it, so either `i18n=""` or `i18n-<attrName>=""`. Something like `$('i18n*')` or a regex selector would have been nice but is not supported.
  * workaround: select all elements from the template and then filter by looking at the attribute names

**Comment by NetanelBasal** [comment]
Thanks for the detailed summary. I'm glad that you found the solutions. If it's not enough and you have more edge cases, it will be safer to use Regex in this case.

**Comment by NetanelBasal** [comment]
@sroettering we've added this. You can now run `ng g @ngneat/transloco:migrate`, and choose Angular.

If you found any issue, let us know.

**Comment by iiceman40** [comment]
@NetanelBasal sounds great! Here's a follow up question: is it planed to provide a migration back to Angular i18n at some point?
